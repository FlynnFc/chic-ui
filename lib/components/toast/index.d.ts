import { toast } from './toastmanager';
export * from './toastmanager';
export default toast;
//# sourceMappingURL=index.d.ts.map