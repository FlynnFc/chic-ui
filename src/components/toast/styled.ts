import styled from 'styled-components';
import { ToastProps } from './toast';
import { theme as typeColors } from 'config/themes';

export const ToastElement = styled.div<ToastProps>`
  background-color: ${(pr) => typeColors[pr.type!].bgColor};
  color: ${(pr) => typeColors[pr.type!].color};
  transition: 0.3s ease;
  position: relative;
  pointer-events: auto;
  overflow: hidden;
  margin: 0 0 6px;
  padding: 30px;
  margin-bottom: 15px;
  max-height: 100px;
  border-radius: 8px;
  opacity: 0.9;
  background-position: 15px;
  background-repeat: no-repeat;
  min-height: 50px;
  max-width: 380px;
  padding: 20px 15px 10px 10px;
  &:hover {
    // box-shadow: 0 0 3px #fff;
    opacity: 1;
    cursor: pointer;
  }
  button {
    position: relative;
    right: -0.3em;
    top: -0.3em;
    float: right;
    font-weight: 700;
    color: ${(pr) => typeColors[pr.type!].color};
    outline: none;
    border: none;
    text-shadow: 0 1px 0 #fff;
    opacity: 0.8;
    line-height: 1;
    font-size: 16px;
    padding: 0;
    cursor: pointer;
    background: 0 0;
    border: 0;
  }
`;

export const ToastTitle = styled.p`
  font-weight: 700;
  font-size: 15px;
  text-align: left;
  margin-top: 0;
  margin-bottom: 6px;
  width: 300px;
  height: 18px;
`;
export const ToastMessage = styled.p`
  margin: 0;
  text-align: left;
  height: 18px;
  margin-left: -1px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;
export const ToastIcon = styled.div`
  float: left;
  margin-right: 15px;
  margin-top: 5px;
  width: 25px;
  height: 25px;
`;
