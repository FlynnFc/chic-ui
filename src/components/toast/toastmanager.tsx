import React from 'react';
import ReactDOM from 'react-dom';
import Toast, { ToastProps } from './toast';
import { themeType } from 'config/themes';

interface ToastOptions {
  id?: string;
  title: string;
  type?: themeType;
  message?: string;
  duration?: number;
  icon?: React.ElementType;
}

export class ToastManager {
  private containerRef: HTMLDivElement;
  private toasts: ToastProps[] = [];

  constructor() {
    const body = document.getElementsByTagName('body')[0] as HTMLBodyElement;
    const toastContainer = document.createElement('div') as HTMLDivElement;
    toastContainer.id = 'main-toast-container';
    toastContainer.style.position = 'fixed';
    toastContainer.style.top = '10px';
    toastContainer.style.right = '10px';
    toastContainer.style.zIndex = '1000';
    body.insertAdjacentElement('beforeend', toastContainer);
    this.containerRef = toastContainer;
  }

  public open(options: ToastOptions): void {
    const toastId = Math.random().toString(36).substring(2, 9);
    const toast: ToastProps = {
      id: toastId,
      // if id is passed within options, it will overwrite the auto-generated one
      ...options,
      close: () => this.close(options.id ?? toastId)
    };

    this.toasts = [toast, ...this.toasts];
    this.render();
  }

  public close(id: string): void {
    this.toasts = this.toasts.filter((toast: ToastProps) => toast.id !== id);
    this.render();
  }

  private render(): void {
    const toastsList = this.toasts.map((toastProps: ToastProps) => (
      <Toast key={toastProps.id} {...toastProps} />
    ));
    ReactDOM.render(toastsList, this.containerRef);
  }
}

export const toast = new ToastManager();
