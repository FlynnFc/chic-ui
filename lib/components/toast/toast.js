var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React, { useEffect } from 'react';
import { ToastElement, ToastTitle, ToastMessage, ToastIcon } from './styled';
var Toast = function (props) {
    var close = props.close, message = props.message, title = props.title, _a = props.duration, duration = _a === void 0 ? 0 : _a, id = props.id, _b = props.type, type = _b === void 0 ? 'primary' : _b, icon = props.icon;
    useEffect(function () {
        if (!duration)
            return;
        var timer = setTimeout(function () {
            close();
        }, duration);
        console.log('RENDERED');
        return function () { return clearTimeout(timer); };
    }, [close, duration]);
    return (_jsxs(ToastElement, __assign({ close: close, message: message, title: title, duration: duration, id: id, type: type, icon: icon }, { children: [_jsx("button", __assign({ onClick: close }, { children: "X" }), void 0), icon && _jsx(ToastIcon, { as: icon }, void 0), _jsxs("div", { children: [_jsx(ToastTitle, { children: title }, void 0), _jsx(ToastMessage, { children: message }, void 0)] }, void 0)] }), void 0));
};
var shouldRerender = function (prevProps, nextProps) {
    return prevProps.id === nextProps.id;
};
export default React.memo(Toast, shouldRerender);
//# sourceMappingURL=toast.js.map