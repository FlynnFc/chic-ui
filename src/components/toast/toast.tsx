import React, { ReactElement } from 'react';
import check from './icons/check.svg';
import error from './icons/error.svg';
import info from './icons/info.svg';
import warning from './icons/warning.svg';
import { ToastElement, ToastImage, ToastMessage, ToastTitle } from './styled';

export interface ToastProps {
  title: string;
  message?: string;
  type: 'success' | 'error' | 'info' | 'warning';
  onClose?: () => void;
}

const toastAlternatives = {
  success: { backgroundColor: '#5cb85c', icon: check },
  error: { backgroundColor: '#d9534f', icon: error },
  info: { backgroundColor: '#5bc0de', icon: info },
  warning: { backgroundColor: '#f0ad4e', icon: warning }
};
function Toast({
  title,
  message = '',
  type,
  onClose = () => {}
}: ToastProps): ReactElement {
  return (
    <ToastElement
      style={{
        backgroundColor: toastAlternatives[type].backgroundColor
      }}
    >
      <button onClick={() => onClose()}>X</button>
      <ToastImage>
        <img src={toastAlternatives[type].icon} alt="" />
      </ToastImage>
      <div>
        <ToastTitle>{title}</ToastTitle>
        <ToastMessage>{message}</ToastMessage>
      </div>
    </ToastElement>
  );
}

export default Toast;
