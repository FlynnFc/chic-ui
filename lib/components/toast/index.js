import { toast } from './toastmanager';
export * from './toastmanager';
export default toast;
//# sourceMappingURL=index.js.map