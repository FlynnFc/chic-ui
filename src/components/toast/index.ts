import { toast } from './toastmanager';

export * from './toastmanager';

export default toast;