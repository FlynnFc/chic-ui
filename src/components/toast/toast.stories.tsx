import { Meta } from '@storybook/react/types-6-0';
import Toast from './toast';
import { toast } from './toastmanager';
import SearchIcon from '../../icons/Search';

export default {
  title: 'Components/Toast',
  component: Toast
} as Meta;

// display a toast message when the page loads
// you can call toast.destroy('my-id') from anywhere to remove this message manually with code

export const Trying = () => {
  return (
    <button
      onClick={() =>
        toast.open({
          title: 'Toast title',
          message: 'Toast body',
          type: 'secondary',
          duration: 3000
        })
      }
    >
      Show toast
    </button>
  );
};
