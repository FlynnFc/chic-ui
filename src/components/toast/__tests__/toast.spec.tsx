import toast from '..';
import { render } from '../../../test/utils';

describe('Toast Component', () => {
  it('Renders Correctly', () => {
    const ToastGroup = render(
      <>
        <button
          onClick={() =>
            toast.open({
              title: 'ERROR!',
              message: 'Please enter your name',
              type: 'danger',
              duration: 3000
            })
          }
        >
          Show toast
        </button>
      </>
    );
    expect(ToastGroup).toMatchSnapshot();
  });
});
