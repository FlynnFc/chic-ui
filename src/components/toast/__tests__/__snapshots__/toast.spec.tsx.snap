// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Component Renders Correctly 1`] = `
<button
  onClick={[Function]}
>
  Show toast
</button>
`;
